cli/src/lib/block-explorer-client.ts

```diff
@@ -95,19 +107,46 @@ export class BlockExplorerClient {
       throw new Error(`Received empty Source Code for ${rawAddress}`);
     }
 
-    const rawSourceCode = result[0].SourceCode.replace(/^\{\{/, "{").replace(/}}$/, "}");
+    if (result[0].SourceCode === "" || result[0].ABI === "Contract source code not verified") {
+      throw new ContracNotVerified(rawAddress);
+    }
+
+    const abi = JSON.parse(result[0].ABI);
+    this.abiCache.set(rawAddress, abi);
+
     try {
+      const rawSourceCode = result[0].SourceCode.replace(/^\{\{/, "{").replace(/}}$/, "}");
       const SourceCode = sourceCodeSchema.parse(JSON.parse(rawSourceCode));
       const data = new ContractData(result[0].ContractName, SourceCode.sources, contractAddr);
       this.sourceCache.set(rawAddress, data);
       return data;
     } catch (e) {
+      // This means that the response was not an object, instead it was a string with the source code.
+      // We cannot recreate the dir structure in this case. We also don't know the right file name or file type.
       if (e instanceof SyntaxError) {
-        const sources = { "main.sol": { content: rawSourceCode } };
-        const data = new ContractData(result[0].ContractName, sources, contractAddr);
+        const content = { content: result[0].SourceCode };
+        const data = new ContractData(
+          result[0].ContractName,
+          { "contract.sol": content },
+          contractAddr
+        );
```

Comment:
i know this error handling sounds terrible. But I'm 97% sure that this is close to the best we can do in this case.

Reply:
This means that we got an Etherscan response that we are not able to parse, and convert into a proper class?
In what case can this happen? Dont we would need to say something?